File: backend/app/routers/crm.py
"""CRM students CRUD (admin tooling). Manager-only per the auth matrix."""
from fastapi import APIRouter, Depends, status
from pymysql.connections import Connection

from ..auth import require_manager
from ..database import get_db
from ..models import StudentCrm

router = APIRouter(prefix="/api/crm/students", tags=["crm"])

_SELECT = (
    "SELECT id, name, phone, course_id AS courseId, course_name AS courseName, "
    "status, progress_score AS progressScore, notes, "
    "DATE_FORMAT(created_at, '%%Y-%%m-%%d') AS createdAt "
    "FROM students_crm ORDER BY created_at DESC, id DESC"
)


@router.get("", response_model=list[StudentCrm])
def list_students(db: Connection = Depends(get_db), _: object = Depends(require_manager)):
    with db.cursor() as cur:
        cur.execute(_SELECT)
        return [StudentCrm(**row) for row in cur.fetchall()]


@router.post("", response_model=StudentCrm, status_code=status.HTTP_201_CREATED)
def create_student(student: StudentCrm, db: Connection = Depends(get_db), _: object = Depends(require_manager)):
    with db.cursor() as cur:
        cur.execute(
            "INSERT INTO students_crm "
            "(name, phone, course_id, course_name, status, progress_score, notes) "
            "VALUES (%(name)s, %(phone)s, %(courseId)s, %(courseName)s, "
            "%(status)s, %(progressScore)s, %(notes)s)",
            student.model_dump(exclude={"id", "createdAt"}),
        )
        student.id = cur.lastrowid
    return student


@router.put("/{student_id}", response_model=StudentCrm)
def update_student(student_id: int, student: StudentCrm, db: Connection = Depends(get_db), _: object = Depends(require_manager)):
    student.id = student_id
    with db.cursor() as cur:
        cur.execute(
            "UPDATE students_crm SET name=%(name)s, phone=%(phone)s, "
            "course_id=%(courseId)s, course_name=%(courseName)s, status=%(status)s, "
            "progress_score=%(progressScore)s, notes=%(notes)s WHERE id=%(id)s",
            student.model_dump(exclude={"createdAt"}),
        )
    return student


@router.delete("/{student_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_student(student_id: int, db: Connection = Depends(get_db), _: object = Depends(require_manager)):
    with db.cursor() as cur:
        cur.execute("DELETE FROM students_crm WHERE id=%(id)s", {"id": student_id})

File: core/handlers/crm.go
package handlers

// CRM students CRUD (admin tooling). Manager-only (gated at the route).
// Migrated from the FastAPI service. Bare JSON bodies.

import (
	"net/http"

	"handayani-core/database"
	"handayani-core/models"

	"github.com/gin-gonic/gin"
)

func ListStudentsCrm(c *gin.Context) {
	students := make([]models.StudentCrm, 0)
	// Project created_at as a date string to match the Python wire format.
	if err := database.DB.Model(&models.StudentCrm{}).
		Select("id, name, phone, course_id, course_name, status, progress_score, notes, " +
			"DATE_FORMAT(created_at, '%Y-%m-%d') AS created_at").
		Order("created_at DESC, id DESC").
		Find(&students).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to list students"})
		return
	}
	c.JSON(http.StatusOK, students)
}

func CreateStudentCrm(c *gin.Context) {
	var student models.StudentCrm
	if err := c.ShouldBindJSON(&student); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if student.Status == "" {
		student.Status = "lead" // mirror the Pydantic default
	}
	student.ID = 0
	student.CreatedAt = nil // let the DB default fill created_at
	// Select only the writable columns so GORM doesn't try to insert created_at.
	if err := database.DB.Select(
		"name", "phone", "course_id", "course_name", "status", "progress_score", "notes",
	).Create(&student).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create student"})
		return
	}
	c.JSON(http.StatusCreated, student)
}

func UpdateStudentCrm(c *gin.Context) {
	id, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid id"})
		return
	}
	var student models.StudentCrm
	if err := c.ShouldBindJSON(&student); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	student.ID = id
	// Update the 7 writable columns; never touch created_at.
	if err := database.DB.Model(&models.StudentCrm{}).Where("id = ?", id).Updates(map[string]interface{}{
		"name":           student.Name,
		"phone":          student.Phone,
		"course_id":      student.CourseID,
		"course_name":    student.CourseName,
		"status":         student.Status,
		"progress_score": student.ProgressScore,
		"notes":          student.Notes,
	}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update student"})
		return
	}
	c.JSON(http.StatusOK, student)
}

func DeleteStudentCrm(c *gin.Context) {
	id, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid id"})
		return
	}
	if err := database.DB.Delete(&models.StudentCrm{}, id).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete student"})
		return
	}
	c.Status(http.StatusNoContent)
}
